src/graph/TemporalMergePointsNode.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright 2022 Robotec.AI
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <graph/NodesCore.hpp>
+#include <RGLFields.hpp>
+
+void TemporalMergePointsNode::setParameters(const std::vector<rgl_field_t>& fields)
+{
+	if (std::find(fields.begin(), fields.end(), RGL_FIELD_DYNAMIC_FORMAT) != fields.end()) {
+		throw InvalidAPIArgument("cannot perform temporal merge on field 'RGL_FIELD_DYNAMIC_FORMAT'");
+	}
+
+	mergedData.clear();
+	width = 0;
+
+	for (auto&& field : fields) {
+		if (!mergedData.contains(field) && !isDummy(field)) {
+			mergedData.insert({field, VArray::create(field)});
+		}
+	}
+}
+
+void TemporalMergePointsNode::validate()
+{
+	input = getValidInput<IPointsNode>();
+
+	// Check input pointcloud is unorganized
+	if (input->getHeight() != 1) {
+		auto msg = "Temporal points merge could be proceed on unorganized pointclouds only";
+		throw InvalidPipeline(msg);
+	}
+
+	// Check input pointcloud has required fields
+	for (const auto& [requiredField, _] : mergedData) {
+		if (!input->hasField(requiredField)) {
+			auto msg = fmt::format("TemporalMergePointsNode input has not required field '{}'",
+			                       toString(requiredField));
+			throw InvalidPipeline(msg);
+		}
+	}
+}
+
+void TemporalMergePointsNode::schedule(cudaStream_t stream)
+{
+	for (const auto& [field, data] : mergedData) {
+		size_t pointCount = input->getPointCount();
+		const auto toMergeData = input->getFieldData(field, stream);
+		data->insertData(toMergeData->getReadPtr(MemLoc::Device), pointCount, width);
+	}
+	width += input->getWidth();
```

Comment:
Data must be merged on the Host side. We are already running short of GPU memory with simulators.

Reply:
Do you think this change is enough?
```suggestion
	for (const auto& [field, data] : mergedData) {
		size_t pointCount = input->getPointCount();
		const auto toMergeData = input->getFieldData(field, stream);
		data->insertData(toMergeData->getReadPtr(MemLoc::Host), pointCount, width);
	}
	width += input->getWidth();
```
Output nodes could use merged data on the GPU, so the data will be transferred anyway.

Reply:
You have just changed the source of the data which will needlessly copy input to the host. It is the destination VArray which must reside on the Host. Currently, I'm not able to provide a solution for this (semi-vacation), but I'm okay with merging as-is, provided that you add me a notion task to fix this.

Reply:
OK :bowing_man: